cmd/ipfs/daemon.go

```diff
@@ -205,6 +209,42 @@ func defaultMux(path string) corehttp.ServeOption {
 	}
 }
 
+// setMemoryLimit a soft memory limit to enforce running the GC more often when
+// we are about to run out.
+// This allows to recoop memory when it's about to run out and cancel the
+// doubled memory footprint most go programs experience, at the cost of more CPU
+// usage in memory tight conditions. This does not increase CPU usage when memory
+// is plenty available, it will use more CPU and continue to run in cases where Kubo
+// would OOM.
+func setMemoryLimit() {
+	// From the STD documentation:
+	// A negative input does not adjust the limit, and allows for retrieval of the currently set memory limit.
+	if currentMemoryLimit := debug.SetMemoryLimit(-1); currentMemoryLimit != math.MaxInt64 {
+		fmt.Printf("GOMEMLIMIT already set to %s, leaving as-is.\n", humanize.IBytes(uint64(currentMemoryLimit)))
+		// only update the memory limit if it wasn't set with GOMEMLIMIT already
+		return
+	}
+
+	// this is a proportional negative-rate increase curve fitted equation to thoses points:
+	// 0GiB   -> 0GiB
+	// 4GiB   -> 0.5GiB
+	// 6GiB   -> 0.75GiB
+	// 12GiB  -> 1GiB
+	// 256GiB -> 2GiB
+	totalMemory := memory.TotalMemory()
```

Comment:
Maybe base this on the resource manager instead? (Perhaps with some factor.)

Reply:
I did thought about that but its meant for libp2p's buffer. I dont think it fits well.

Reply:
If anything, I think we should scale go-libp2p based on `GOMEMLIMIT` if it is set.